File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerWorkforceContributionsV2.kt
package com.amaury.pointage.v2.engine

import java.time.YearMonth
import kotlin.math.min

/**
 * Cotisations patronales liées à l'effectif en 2026.
 *
 * La tranche d'effectif ne suffit pas à elle seule à déterminer le FNAL pour tous les employeurs :
 * certains employeurs agricoles/cooperatifs relèvent du taux plafonné de 0,10 % même avec un
 * effectif d'au moins 50 salariés. L'applicabilité de la contribution formation peut elle aussi
 * dépendre de la situation du salarié (ex. exonération confirmée).
 *
 * HoraTrack exige donc ces faits explicitement au lieu de les déduire silencieusement.
 */
object EmployerWorkforceContributionsV2 {
    enum class Band { UNDER_11, FROM_11_TO_49, AT_LEAST_50 }

    enum class FnalTreatment {
        /** 0,10 % dans la limite du plafond de sécurité sociale. */
        CAPPED_0_1_PERCENT,
        /** 0,50 % sur la totalité de l'assiette. */
        UNCAPPED_0_5_PERCENT
    }

    enum class TrainingTreatment {
        /** Contribution légale calculée selon la tranche <11 / >=11. */
        STANDARD,
        /** Exonération/applicabilité nulle confirmée pour la rémunération considérée. */
        EXEMPT_CONFIRMED
    }

    data class Record(
        val id: String,
        val band: Band,
        val effectiveFrom: YearMonth,
        val effectiveTo: YearMonth? = null,
        val source: String,
        /** Régime FNAL confirmé ; null reste inconnu pour une migration fail-closed. */
        val fnalTreatment: FnalTreatment? = null,
        /** Applicabilité formation confirmée ; null reste inconnu pour une migration fail-closed. */
        val trainingTreatment: TrainingTreatment? = null
    )

    data class Snapshot(
        val band: Band?,
        val source: String?,
        val reliable: Boolean,
        val warnings: List<String>,
        val fnalTreatment: FnalTreatment? = null,
        val trainingTreatment: TrainingTreatment? = null
    )

    data class Result(
        val fnalAmount: Double?,
        val trainingAmount: Double?,
        val totalEmployerAmount: Double?,
        val complete: Boolean,
        val warnings: List<String>
    )

    fun resolve(records: List<Record>, period: YearMonth): Snapshot {
        val malformed = records.filter {
            it.source.isBlank() || it.effectiveTo?.let { end -> end < it.effectiveFrom } == true
        }
        if (malformed.isNotEmpty()) {
            return Snapshot(
                band = null,
                source = null,
                reliable = false,
                warnings = listOf("Effectif employeur : une règle enregistrée est incomplète ou incohérente ; FNAL/formation non calculés."),
                fnalTreatment = null,
                trainingTreatment = null
            )
        }
        val active = records.filter {
            period >= it.effectiveFrom && (it.effectiveTo == null || period <= it.effectiveTo)
        }
        if (active.isEmpty()) {
            return Snapshot(
                band = null,
                source = null,
                reliable = false,
                warnings = listOf("Effectif employeur : tranche <11 / 11–49 / >=50, régime FNAL et applicabilité formation à confirmer pour ${period.monthValue.toString().padStart(2, '0')}/${period.year} ; cotisations incomplètes."),
                fnalTreatment = null,
                trainingTreatment = null
            )
        }
        if (active.size > 1) {
            return Snapshot(
                band = null,
                source = null,
                reliable = false,
                warnings = listOf("Effectif employeur : plusieurs règles se chevauchent sur la période ; FNAL/formation bloqués."),
                fnalTreatment = null,
                trainingTreatment = null
            )
        }
        val selected = active.single()
        val warnings = buildList {
            if (selected.fnalTreatment == null) {
                add("FNAL : régime 0,10 % plafonné / 0,50 % déplafonné à confirmer ; la tranche d'effectif seule ne suffit pas pour tous les employeurs.")
            }
            if (selected.trainingTreatment == null) {
                add("Formation professionnelle : applicabilité/exonération à confirmer pour la rémunération considérée.")
            }
        }
        return Snapshot(
            band = selected.band,
            source = selected.source,
            reliable = selected.fnalTreatment != null && selected.trainingTreatment != null,
            warnings = warnings,
            fnalTreatment = selected.fnalTreatment,
            trainingTreatment = selected.trainingTreatment
        )
    }

    fun calculate(
        grossSocial: Double,
        applicableMonthlyCeiling: Double?,
        year: Int,
        band: Band?,
        fnalTreatment: FnalTreatment? = null,
        trainingTreatment: TrainingTreatment? = null
    ): Result {
        if (year != 2026) {
            return Result(null, null, null, false, listOf("FNAL/formation : barème non intégré pour $year."))
        }
        if (!grossSocial.isFinite() || grossSocial < 0.0) {
            return Result(null, null, null, false, listOf("FNAL/formation : assiette brute sociale invalide ; aucun montant patronal n'est calculé."))
        }

        val warnings = mutableListOf<String>()

        val fnal = when (fnalTreatment) {
            null -> {
                warnings += "FNAL : régime 0,10 % plafonné / 0,50 % déplafonné à confirmer ; aucun taux n'est déduit du seul effectif."
                null
            }
            FnalTreatment.CAPPED_0_1_PERCENT -> {
                val ceiling = applicableMonthlyCeiling
                if (ceiling == null || !ceiling.isFinite() || ceiling < 0.0) {
                    warnings += "FNAL : plafond social applicable indisponible pour le régime plafonné."
                    null
                } else {
                    min(grossSocial, ceiling) * 0.0010
                }
            }
            FnalTreatment.UNCAPPED_0_5_PERCENT -> grossSocial * 0.0050
        }

        val training = when (trainingTreatment) {
            null -> {
                warnings += "Formation professionnelle : applicabilité/exonération à confirmer pour la rémunération considérée."
                null
            }
            TrainingTreatment.EXEMPT_CONFIRMED -> 0.0
            TrainingTreatment.STANDARD -> {
                when (band) {
                    null -> {
                        warnings += "Formation professionnelle : tranche d'effectif <11 / >=11 à confirmer."
                        null
                    }
                    Band.UNDER_11 -> grossSocial * 0.0055
                    Band.FROM_11_TO_49, Band.AT_LEAST_50 -> grossSocial * 0.0100
                }
            }
        }

        val complete = fnal != null && training != null
        return Result(
            fnalAmount = fnal,
            trainingAmount = training,
            totalEmployerAmount = if (complete) fnal!! + training!! else null,
            complete = complete,
            warnings = warnings.distinct()
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/EmployerWorkforceContributionsV2.swift
import Foundation

/// FNAL et formation professionnelle 2026 selon la tranche d'effectif social confirmée.
enum EmployerWorkforceContributionsV2 {
    enum Band: Equatable {
        case under11
        case from11To49
        case atLeast50
    }

    struct Period: Equatable, Comparable {
        let year: Int
        let month: Int

        var isValid: Bool { year > 0 && (1...12).contains(month) }

        static func < (lhs: Period, rhs: Period) -> Bool {
            lhs.year == rhs.year ? lhs.month < rhs.month : lhs.year < rhs.year
        }

        var label: String { String(format: "%02d/%04d", month, year) }
    }

    struct Record: Equatable {
        let id: String
        let band: Band
        let effectiveFrom: Period
        let effectiveTo: Period?
        let source: String
    }

    struct Snapshot: Equatable {
        let band: Band?
        let source: String?
        let reliable: Bool
        let warnings: [String]
    }

    struct Result: Equatable {
        let fnalAmount: Double?
        let trainingAmount: Double?
        let totalEmployerAmount: Double?
        let complete: Bool
        let warnings: [String]
    }

    static func resolve(records: [Record], period: Period) -> Snapshot {
        guard period.isValid else {
            return Snapshot(
                band: nil,
                source: nil,
                reliable: false,
                warnings: ["Effectif employeur : période invalide ; FNAL/formation non calculés."]
            )
        }

        let malformed = records.filter { record in
            record.source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                !record.effectiveFrom.isValid ||
                (record.effectiveTo?.isValid == false) ||
                (record.effectiveTo.map { $0 < record.effectiveFrom } ?? false)
        }
        guard malformed.isEmpty else {
            return Snapshot(
                band: nil,
                source: nil,
                reliable: false,
                warnings: ["Effectif employeur : une règle enregistrée est incomplète ou incohérente ; FNAL/formation non calculés."]
            )
        }

        let active = records.filter { record in
            period >= record.effectiveFrom && (record.effectiveTo == nil || period <= record.effectiveTo!)
        }
        guard !active.isEmpty else {
            return Snapshot(
                band: nil,
                source: nil,
                reliable: false,
                warnings: ["Effectif employeur : tranche <11 / 11–49 / ≥50 à confirmer pour \(period.label) ; FNAL/formation incomplets."]
            )
        }
        guard active.count == 1, let selected = active.first else {
            return Snapshot(
                band: nil,
                source: nil,
                reliable: false,
                warnings: ["Effectif employeur : plusieurs tranches se chevauchent sur la période ; FNAL/formation bloqués."]
            )
        }

        return Snapshot(
            band: selected.band,
            source: selected.source,
            reliable: true,
            warnings: []
        )
    }

    static func calculate(
        grossSocial: Double,
        applicableMonthlyCeiling: Double?,
        year: Int,
        band: Band?
    ) -> Result {
        guard year == 2026 else {
            return incomplete("FNAL/formation : barème non intégré pour \(year).")
        }
        guard let band else {
            return incomplete("FNAL/formation : tranche d'effectif à confirmer ; coût employeur incomplet.")
        }
        guard let ceiling = applicableMonthlyCeiling,
              ceiling.isFinite,
              ceiling >= 0 else {
            return incomplete("FNAL/formation : plafond social applicable indisponible.")
        }
        guard grossSocial.isFinite, grossSocial >= 0 else {
            return incomplete("FNAL/formation : assiette brute sociale invalide ; aucun montant patronal n'est calculé.")
        }

        let fnalBase: Double
        let fnalRate: Double
        let trainingRate: Double
        switch band {
        case .under11:
            fnalBase = min(grossSocial, ceiling)
            fnalRate = 0.0010
            trainingRate = 0.0055
        case .from11To49:
            fnalBase = min(grossSocial, ceiling)
            fnalRate = 0.0010
            trainingRate = 0.0100
        case .atLeast50:
            fnalBase = grossSocial
            fnalRate = 0.0050
            trainingRate = 0.0100
        }

        let fnal = fnalBase * fnalRate
        let training = grossSocial * trainingRate
        return Result(
            fnalAmount: fnal,
            trainingAmount: training,
            totalEmployerAmount: fnal + training,
            complete: true,
            warnings: []
        )
    }

    private static func incomplete(_ warning: String) -> Result {
        Result(
            fnalAmount: nil,
            trainingAmount: nil,
            totalEmployerAmount: nil,
            complete: false,
            warnings: [warning]
        )
    }
}
